proc-macros/src/render_client.rs

```diff
@@ -68,6 +69,46 @@ impl RpcDescription {
 		Ok(trait_impl)
 	}
 
+	/// Verify and rewrite the return type (for methods).
+	fn return_result_type(&self, mut ty: syn::Type) -> TokenStream2 {
+		// We expect a valid type path.
+		let syn::Type::Path(ref mut type_path) = ty else  {
+			return quote_spanned!(ty.span() => compile_error!("Expecting something like 'Result<Foo, Err>' here. (1)"));
+		};
+
+		// The path (eg std::result::Result) should have a final segment like 'Result'.
+		let Some(type_name) = type_path.path.segments.last_mut() else {
+			return quote_spanned!(ty.span() => compile_error!("Expecting this path to end in something like 'Result<Foo, Err>'"));
+		};
+
+		// Get the generic args eg the <T, E> in Result<T, E>.
+		let PathArguments::AngleBracketed(AngleBracketedGenericArguments { args, .. }) = &mut type_name.arguments else {
+			return quote_spanned!(ty.span() => compile_error!("Expecting something like 'Result<Foo, Err>' here, but got no generic args (eg no '<Foo,Err>')."));
+		};
+
+		if type_name.ident == "Result" {
```

Comment:
technically I'm not sure that we need these guards anymore as long the error type impls Into<ErrorObjectOwned>

one would get some weird compile error once it doesn't I guess.

Reply:
The only reason we have these guards etc now is to check what sort of type we're looking at in order to trust that the `T` param is the return type so that we can rewrite the error properly to a `Result<T, core::Error>`.

Without them things would work still but be more permissive and may lead to weird compile errors with custom types rather than our nicer "use a Result or RpcResult" type ones.

Reply:
gotcha